### PR TITLE
ENH: Warn about versions in sys_info

### LINKED ISSUE #12132: DOC: Actively suggest updating MNE to a supported version
I often have people come to me locally with bugs, only to find that they are on some old version of MNE-Python. I think it's not generally known to neuroscience people that upgrading is important for bug fixes etc., so it would be good to make it clear what the software support policy is and that upgrading is important.

We should document this clearly as much as possible, but I have two other ideas that will actively "pop up" / ping users more actively:

1. Add a date-based warning *on import*. For example, if we release 1.6 on Nov 1 2023, add to `mne/__init__.py` code to emit a warning on Nov 1 2024 that a new version of MNE is available and that they should update to avoid bugs (or some similar explicit message). We don't absolutely know or guarantee our release cadence, but the longest we have gone between releases in the last few years is probably like ~8 mo, and I think we now shoot for ~4 mo -- so to be safe if can make it year from when we cut a release. This could be controllable with a `MNE_CHECK_VERSION_DATE` env var or something in case people have wacky datetime settings.
2. Query the current release version over the internet when `mne.sys_info` is called with a new kwarg `check_version=True`. We don't want our software to start trying to "phone home" without a user asking in some way (i.e., definitely *not* at import time). Then the question becomes what should the default value of `check_version` in `mne.sys_info(..., check_version=...)` be:

   - `False`: least invasive but also least informative for users. In our bug report forms and forum stuff we ask people to call `mne sys_info --check-version` or `mne.sys_info(check_version=True)` or so. But the probability of users seeing an informative warning goes down.
   - `True`: most invasive but also most informative for users.
   - `None`: in between if we make it mean "try to check and if it takes more than 2 sec don't bother" or something. (FWIW, `mne sys_info` on my machine currently takes ~3.7s.)

   Whichever way we go, the default behavior for `None` could be controllable with `MNE_SYS_INFO_CHECK_VERSION` var in case people want the default `None` behavior to mean `False` or `True`.

Happy to hear thoughts on these and other ideas for suggesting people update to supported versions.

### PR DESCRIPTION
**Pull request by larsoner**
This is the best formatting I could find:

1. Add version check right by where the `mne` version is reported, *and*
2. If outdated, add a line at the end gently suggesting to update

With this change I get:

## On dev

On my machine I get:
```
$ mne sys_info
Platform             Linux-6.2.0-35-generic-x86_64-with-glibc2.37
Python               3.11.4 (main, Jun  9 2023, 07:59:55) [GCC 12.3.0]
Executable           /home/larsoner/python/virtualenvs/base/bin/python3
CPU                  x86_64 (8 cores)
Memory               62.7 GB

Core
├☑ mne               1.6.0.dev154+ga320b631a7.d20231024
│                    ✓ Dev version (latest release: 1.5.1)
├☑ numpy             2.0.0.dev0+git20231025.9f6789c (OpenBLAS 0.3.23.dev with 4 threads)
├☑ scipy             1.12.0.dev0+1888.d7f12d3
...
├☑ mne-bids-pipeline 1.4.0
└☐ unavailable       mne-nirs, mne-features, mne-connectivity, mne-icalabel
```
## On stable

Hacking the code to say I'm on stable the above changes to:
```
├☑ mne               1.5.1
│                    ✓ On the latest release
```

## Outdated

Hacking the code to fake the latest release to be 1.7 we see the two lines change:
```
$ mne sys_info
Platform             Linux-6.2.0-35-generic-x86_64-with-glibc2.37
Python               3.11.4 (main, Jun  9 2023, 07:59:55) [GCC 12.3.0]
Executable           /home/larsoner/python/virtualenvs/base/bin/python3
CPU                  x86_64 (8 cores)
Memory               62.7 GB

Core
├☑ mne               1.6.0.dev154+ga320b631a7.d20231024
│                    ✗ A newer version 1.7 is available!
├☑ numpy             2.0.0.dev0+git20231025.9f6789c (OpenBLAS 0.3.23.dev with 4 threads)
├☑ scipy             1.12.0.dev0+1888.d7f12d3
...
├☑ mne-bids-pipeline 1.4.0
└☐ unavailable       mne-nirs, mne-features, mne-connectivity, mne-icalabel

To update to the latest supported release version to get bugfixes and improvements, visit https://mne.tools/stable/install/updating.html
```

## Timeout (default 2s)

If I hack the timeout to be tiny enough that it fails for my system, same 

```
├☑ mne               1.6.0.dev154+ga320b631a7.d20231024
│                    Unable to check latest version on GitHub (2.0 sec timeout)
```

## Todo

Just waiting for feedback before finishing:

- [ ] Tests
- [ ] Changelog entry

Closes #12132

### CODE REVIEW

File: mne/utils/config.py [hunk 1]
```diff
@@ -17,6 +17,10 @@
 from functools import partial
 from importlib import import_module
 from pathlib import Path
+from urllib.error import URLError
+from urllib.request import urlopen
+
+from packaging.version import parse
```
Comment (drammock) [inline mne/utils/config.py:23]: I'm surprised this passed style checks, shouldn't it fail on import sorting grounds?

**Review by drammock — APPROVED** [review]
thanks for doing this!  My (bikeshedding) nitpicks are:

- prefer to have the message on the same line as the reported version (similar to how we put BLAS info in parentheses after reporting numpy version)
- instead of an extra check/x-mark in the message, I would consider having the main left-margin checkbox be the indicator, like
    - ✓ (installed stable or dev)
    - x (installed old version)

File: mne/utils/config.py (hunk 1 above)
Comment (larsoner) [reply mne/utils/config.py:23]: This makes it a different import block

**Comment by larsoner** [comment]
> prefer to have the message on the same line as the reported version (similar to how we put BLAS info in parentheses after reporting numpy version)

Actually this split is intentional. It gives people two chances to see the version mismatch. If you really want them both places, I guess I'd prefer and can live with them both in a seprate message at the bottom since it's more visible than when it's up by `mne`.

**Comment by drammock** [comment]
> > prefer to have the message on the same line as the reported version (similar to how we put BLAS info in parentheses after reporting numpy version)
> 
> Actually this split is intentional. It gives people two chances to see the version mismatch. If you really want them _EDIT: in one place_, I guess I'd prefer and can live with them both in a separate message at the bottom since it's more visible than when it's up by `mne`.

I wasn't being clear.  I like that you add a warning-like sentence after *the whole results table is done*.  What I meant was that I don't like the entry for MNE spanning two lines.

**Comment by larsoner** [comment]
So something like:
```
Platform             Linux-6.2.0-35-generic-x86_64-with-glibc2.37
Python               3.11.4 (main, Jun  9 2023, 07:59:55) [GCC 12.3.0]
Executable           /home/larsoner/python/virtualenvs/base/bin/python3
CPU                  x86_64 (8 cores)
Memory               62.7 GB

Core
├☒ mne               1.6.0.dev154+ga320b631a7.d20231024 (outdated, release 1.7.0 is available!)
├☑ numpy             2.0.0.dev0+git20231025.9f6789c (OpenBLAS 0.3.23.dev with 4 threads)
...
├☑ mne-bids          0.14.dev0
├☑ mne-bids-pipeline 1.4.0
└☐ unavailable       mne-nirs, mne-features, mne-connectivity, mne-icalabel

To update to the latest supported release version to get bugfixes and improvements, visit https://mne.tools/stable/install/updating.html
```
and for the other variants:
```
├☑ mne               1.6.0.dev154+ga320b631a7.d20231024 (devel, latest release is 1.5.1)

├☑ mne               1.5.1 (latest release)

├☑ mne               1.6.0.dev154+ga320b631a7.d20231024 (unable to check for latest version on GitHub, 2.0 sec timeout)
```

**Comment by drammock** [comment]
> So something like:

yes exactly!

**Comment by larsoner** [comment]
Okay pushed a commit for that. Just need to add a couple tests and changelog!

**Comment by larsoner** [comment]
Okay tests and changelog updated

File: mne/utils/config.py [hunk 2]
```diff
@@ -719,7 +739,16 @@ def sys_info(fid=None, show_paths=False, *, dependencies="user", unicode=True):
...
+            mark = "☑" if unicode else "+"
+            mne_extra = ""
+            if mod_name == "mne" and check_version:
+                timeout = 2.0 if check_version is True else float(check_version)
+                mne_version_good, mne_extra = _check_mne_version(timeout)
+                if mne_version_good is None:
+                    mne_version_good = True
+                elif not mne_version_good:
+                    mark = "☒" if unicode else "X"
+            out(f"{pre}{mark} " if unicode else " {mark} ")
```
Comment (drammock) [inline mne/utils/config.py:751]:
```suggestion
            out(f"{pre}{mark} " if unicode else f" {mark} ")
```

I think you need the `f` in both places, no?

**Review by drammock — COMMENTED** [review]
LGTM except for one oversight I think, see below

**Comment by hoechenberger** [comment]
We could also query PyPI to ensure the package is already pip-installable by the time we display the message:
https://pypi.python.org/pypi/mne/json

**Comment by hoechenberger** [comment]
@larsoner
Something to consider before merging

I'm behind a corporate firewall that injects a self-signed MITM certificate so it can spy on TLS connections

Which can cause troubles connecting to online services unless one explicitly specifies that one trusts this cert

For example, for Python's `requests` I have to set the `REQUESTS_CA_BUNDLE` env var or else connections may fail.

I'd like to try out how this branch behaves in different settings (with and without the env var set; connection from the office and remotely through a VPN) before we merge, just to ensure sys_info doesn't suddenly explode or spit out many lines of error messages

But I won't have time to test before next week

**Comment by drammock** [comment]
> I'd like to try out how this branch behaves in different settings (with and without the env var set; connection from the office and remotely through a VPN; on the host machine and inside a Docker Dev Container) before we merge

Good idea, IMO there is no rush on this one

**Comment by hoechenberger** [comment]
The exception that occurs is:

```
URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1006)>
```

**Comment by hoechenberger** [comment]
… and the solution is to set the `SSL_CERT_FILE` env var to the path of the self-signed root certificate.

File: mne/utils/config.py [hunk 3]
```diff
@@ -756,3 +788,38 @@ def sys_info(fid=None, show_paths=False, *, dependencies="user", unicode=True):
...
+        return None
+    else:
+        return response["tag_name"]
+
+
+def _check_mne_version(timeout):
+    rel_ver = _get_latest_version(timeout)
+    if rel_ver is None:
+        return None, (
+            f"unable to check for latest version on GitHub, {timeout} sec timeout"
```
Comment (larsoner) [inline mne/utils/config.py:816]:
```suggestion
            "unable to check for latest version on GitHub"
```

**Comment by larsoner** [comment]
> So maybe we should just remove the part of the logging message that mentions the timeout?

Now I triage timeouts, SSL errors, and unknown errors. It seemed reasonable to add some explanation to these cases

**Comment by hoechenberger** [comment]
@larsoner Works! :)

**Review by drammock — APPROVED** [review]

**Pull request merged**